### ISSUE TITLE
`IllegalArgumentException("unknown HashAlgorithm")` when server sends Certificate Request with Signature Hash Algorithm which is not supported by bc-java

### ISSUE THREAD

**Issue by mstyura** [bcgit/bc-java#422]
`IllegalArgumentException("unknown HashAlgorithm")` is thrown when server sends `Certificate Request` during handshake with `Signature Hash Algorithm` which is not supported by `bc-java`.

I'm [trying](https://github.com/jitsi/libjitsi/issues/441) to add `DTLS1.2` support to [Jitsi](https://github.com/jitsi/). `Jitsi` uses `bc-java` to implement `DTLS`.
With fixes I'm working on, when `Jitsi` is server during `DTLS` connection establishment everything works, but when `Jitsi` is client during `DTLS` connection establishment -  the connection could not be established due to internal error reported during handshake.

An exception happen with following stack trace:
```
createHash:-1, TlsUtils (org.bouncycastle.crypto.tls)
checkTrackingHash:-1, DeferredHash (org.bouncycastle.crypto.tls)
trackHashAlgorithm:-1, DeferredHash (org.bouncycastle.crypto.tls)
trackHashAlgorithms:-1, TlsUtils (org.bouncycastle.crypto.tls)
clientHandshake:-1, DTLSClientProtocol (org.bouncycastle.crypto.tls)
connect:-1, DTLSClientProtocol (org.bouncycastle.crypto.tls)
runInConnectThread:990, DtlsPacketTransformer (org.jitsi.impl.neomedia.transform.dtls)
access$100:40, DtlsPacketTransformer (org.jitsi.impl.neomedia.transform.dtls)
run:1256, DtlsPacketTransformer$2 (org.jitsi.impl.neomedia.transform.dtls)
```

`createHash` is invoked with value `8`, which is unknown hash algorithm to `bc-java`.

It comes from list of supported `Signature Hash Algorithm` inside `Server`s `Certificate Request` message.

Below I'm attaching logs from [`Wireshark`](https://www.wireshark.org/).

`Client Hello` which is sent by `Jitsi` to `WebRTC` client:
```
22	0.150218950	172.30.220.10	172.30.220.10	DTLSv1.2	202	Client Hello

Frame 22: 202 bytes on wire (1616 bits), 202 bytes captured (1616 bits) on interface 1
    Interface id: 1 (any)
        Interface name: any
    Encapsulation type: Linux cooked-mode capture (25)
    Arrival Time: Oct 21, 2018 01:00:50.903203108 +03
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1540072850.903203108 seconds
    [Time delta from previous captured frame: 0.006163916 seconds]
    [Time delta from previous displayed frame: 0.000000000 seconds]
    [Time since reference or first frame: 0.150218950 seconds]
    Frame Number: 22
    Frame Length: 202 bytes (1616 bits)
    Capture Length: 202 bytes (1616 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: sll:ethertype:ip:udp:dtls]
    [Coloring Rule Name: UDP]
    [Coloring Rule String: udp]
Linux cooked capture
    Packet type: Unicast to us (0)
    Link-layer address type: 772
    Link-layer address length: 6
    Source: 00:00:00_00:00:00 (00:00:00:00:00:00)
    Unused: 0000
    Protocol: IPv4 (0x0800)
Internet Protocol Version 4, Src: 172.30.220.10, Dst: 172.30.220.10
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
        0000 00.. = Differentiated Services Codepoint: Default (0)
        .... ..00 = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    Total Length: 186
    Identification: 0x242e (9262)
    Flags: 0x4000, Don't fragment
        0... .... .... .... = Reserved bit: Not set
        .1.. .... .... .... = Don't fragment: Set
        ..0. .... .... .... = More fragments: Not set
        ...0 0000 0000 0000 = Fragment offset: 0
    Time to live: 64
    Protocol: UDP (17)
    Header checksum: 0x05b3 [validation disabled]
    [Header checksum status: Unverified]
    Source: 172.30.220.10
    Destination: 172.30.220.10
User Datagram Protocol, Src Port: 10000, Dst Port: 54646
    Source Port: 10000
    Destination Port: 54646
    Length: 166
    Checksum: 0x110a [unverified]
    [Checksum Status: Unverified]
    [Stream index: 0]
Datagram Transport Layer Security
    DTLSv1.2 Record Layer: Handshake Protocol: Client Hello
        Content Type: Handshake (22)
        Version: DTLS 1.0 (0xfeff)
        Epoch: 0
        Sequence Number: 0
        Length: 145
        Handshake Protocol: Client Hello
            Handshake Type: Client Hello (1)
            Length: 133
            Message Sequence: 0
            Fragment Offset: 0
            Fragment Length: 133
            Version: DTLS 1.2 (0xfefd)
            Random: 5d6d2fd359eb4b5a77989eb2a792dd5e5f36953c716664a0...
                GMT Unix Time: Sep  2, 2019 18:05:55.000000000 +03
                Random Bytes: 59eb4b5a77989eb2a792dd5e5f36953c716664a0ac6ad880...
            Session ID Length: 0
            Cookie Length: 0
            Cipher Suites Length: 26
            Cipher Suites (13 suites)
                Cipher Suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 (0xc02b)
                Cipher Suite: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA256 (0xc023)
                Cipher Suite: TLS_ECDHE_ECDSA_WITH_AES_128_CBC_SHA (0xc009)
                Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256 (0xc02f)
                Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256 (0xc027)
                Cipher Suite: TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA (0xc013)
                Cipher Suite: TLS_DHE_DSS_WITH_AES_128_GCM_SHA256 (0x00a2)
                Cipher Suite: TLS_DHE_DSS_WITH_AES_128_CBC_SHA256 (0x0040)
                Cipher Suite: TLS_DHE_DSS_WITH_AES_128_CBC_SHA (0x0032)
                Cipher Suite: TLS_DHE_RSA_WITH_AES_128_GCM_SHA256 (0x009e)
                Cipher Suite: TLS_DHE_RSA_WITH_AES_128_CBC_SHA256 (0x0067)
                Cipher Suite: TLS_DHE_RSA_WITH_AES_128_CBC_SHA (0x0033)
                Cipher Suite: TLS_EMPTY_RENEGOTIATION_INFO_SCSV (0x00ff)
            Compression Methods Length: 1
            Compression Methods (1 method)
                Compression Method: null (0)
            Extensions Length: 65
            Extension: supported_groups (len=6)
                Type: supported_groups (10)
                Length: 6
                Supported Groups List Length: 4
                Supported Groups (2 groups)
                    Supported Group: secp256r1 (0x0017)
                    Supported Group: secp384r1 (0x0018)
            Extension: use_srtp (len=7)
                Type: use_srtp (14)
                Length: 7
                SRTP Protection Profiles Length: 4
                SRTP Protection Profile: SRTP_AES128_CM_HMAC_SHA1_80 (0x0001)
                SRTP Protection Profile: SRTP_AES128_CM_HMAC_SHA1_32 (0x0002)
                MKI Length: 0
            Extension: signature_algorithms (len=32)
                Type: signature_algorithms (13)
                Length: 32
                Signature Hash Algorithms Length: 30
                Signature Hash Algorithms (15 algorithms)
                    Signature Algorithm: rsa_pkcs1_sha1 (0x0201)
                        Signature Hash Algorithm Hash: SHA1 (2)
                        Signature Hash Algorithm Signature: RSA (1)
                    Signature Algorithm: SHA224 RSA (0x0301)
                        Signature Hash Algorithm Hash: SHA224 (3)
                        Signature Hash Algorithm Signature: RSA (1)
                    Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
                        Signature Hash Algorithm Hash: SHA256 (4)
                        Signature Hash Algorithm Signature: RSA (1)
                    Signature Algorithm: rsa_pkcs1_sha384 (0x0501)
                        Signature Hash Algorithm Hash: SHA384 (5)
                        Signature Hash Algorithm Signature: RSA (1)
                    Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                        Signature Hash Algorithm Hash: SHA512 (6)
                        Signature Hash Algorithm Signature: RSA (1)
                    Signature Algorithm: SHA1 DSA (0x0202)
                        Signature Hash Algorithm Hash: SHA1 (2)
                        Signature Hash Algorithm Signature: DSA (2)
                    Signature Algorithm: SHA224 DSA (0x0302)
                        Signature Hash Algorithm Hash: SHA224 (3)
                        Signature Hash Algorithm Signature: DSA (2)
                    Signature Algorithm: SHA256 DSA (0x0402)
                        Signature Hash Algorithm Hash: SHA256 (4)
                        Signature Hash Algorithm Signature: DSA (2)
                    Signature Algorithm: SHA384 DSA (0x0502)
                        Signature Hash Algorithm Hash: SHA384 (5)
                        Signature Hash Algorithm Signature: DSA (2)
                    Signature Algorithm: SHA512 DSA (0x0602)
                        Signature Hash Algorithm Hash: SHA512 (6)
                        Signature Hash Algorithm Signature: DSA (2)
                    Signature Algorithm: ecdsa_sha1 (0x0203)
                        Signature Hash Algorithm Hash: SHA1 (2)
                        Signature Hash Algorithm Signature: ECDSA (3)
                    Signature Algorithm: SHA224 ECDSA (0x0303)
                        Signature Hash Algorithm Hash: SHA224 (3)
                        Signature Hash Algorithm Signature: ECDSA (3)
                    Signature Algorithm: ecdsa_secp256r1_sha256 (0x0403)
                        Signature Hash Algorithm Hash: SHA256 (4)
                        Signature Hash Algorithm Signature: ECDSA (3)
                    Signature Algorithm: ecdsa_secp384r1_sha384 (0x0503)
                        Signature Hash Algorithm Hash: SHA384 (5)
                        Signature Hash Algorithm Signature: ECDSA (3)
                    Signature Algorithm: ecdsa_secp521r1_sha512 (0x0603)
                        Signature Hash Algorithm Hash: SHA512 (6)
                        Signature Hash Algorithm Signature: ECDSA (3)
            Extension: ec_point_formats (len=4)
                Type: ec_point_formats (11)
                Length: 4
                EC point formats Length: 3
                Elliptic curves point formats (3)
                    EC point format: uncompressed (0)
                    EC point format: ansiX962_compressed_prime (1)
                    EC point format: ansiX962_compressed_char2 (2)
```

`Server Hello` message which is sent by `WebRTC` client to `Jitsi`:
```
23	0.150667255	172.30.220.10	172.30.220.10	DTLSv1.2	718	Server Hello, Certificate, Server Key Exchange, Certificate Request, Server Hello Done

Frame 23: 718 bytes on wire (5744 bits), 718 bytes captured (5744 bits) on interface 1
    Interface id: 1 (any)
        Interface name: any
    Encapsulation type: Linux cooked-mode capture (25)
    Arrival Time: Oct 21, 2018 01:00:50.903651413 +03
    [Time shift for this packet: 0.000000000 seconds]
    Epoch Time: 1540072850.903651413 seconds
    [Time delta from previous captured frame: 0.000448305 seconds]
    [Time delta from previous displayed frame: 0.000448305 seconds]
    [Time since reference or first frame: 0.150667255 seconds]
    Frame Number: 23
    Frame Length: 718 bytes (5744 bits)
    Capture Length: 718 bytes (5744 bits)
    [Frame is marked: False]
    [Frame is ignored: False]
    [Protocols in frame: sll:ethertype:ip:udp:dtls:x509sat:x509sat:pkcs-1]
    [Coloring Rule Name: UDP]
    [Coloring Rule String: udp]
Linux cooked capture
    Packet type: Unicast to us (0)
    Link-layer address type: 772
    Link-layer address length: 6
    Source: 00:00:00_00:00:00 (00:00:00:00:00:00)
    Unused: 0000
    Protocol: IPv4 (0x0800)
Internet Protocol Version 4, Src: 172.30.220.10, Dst: 172.30.220.10
    0100 .... = Version: 4
    .... 0101 = Header Length: 20 bytes (5)
    Differentiated Services Field: 0x00 (DSCP: CS0, ECN: Not-ECT)
        0000 00.. = Differentiated Services Codepoint: Default (0)
        .... ..00 = Explicit Congestion Notification: Not ECN-Capable Transport (0)
    Total Length: 702
    Identification: 0x242f (9263)
    Flags: 0x4000, Don't fragment
        0... .... .... .... = Reserved bit: Not set
        .1.. .... .... .... = Don't fragment: Set
        ..0. .... .... .... = More fragments: Not set
        ...0 0000 0000 0000 = Fragment offset: 0
    Time to live: 64
    Protocol: UDP (17)
    Header checksum: 0x03ae [validation disabled]
    [Header checksum status: Unverified]
    Source: 172.30.220.10
    Destination: 172.30.220.10
User Datagram Protocol, Src Port: 54646, Dst Port: 10000
    Source Port: 54646
    Destination Port: 10000
    Length: 682
    Checksum: 0x130e [unverified]
    [Checksum Status: Unverified]
    [Stream index: 0]
Datagram Transport Layer Security
    DTLSv1.2 Record Layer: Handshake Protocol: Server Hello
        Content Type: Handshake (22)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 0
        Sequence Number: 0
        Length: 104
        Handshake Protocol: Server Hello
            Handshake Type: Server Hello (2)
            Length: 92
            Message Sequence: 0
            Fragment Offset: 0
            Fragment Length: 92
            Version: DTLS 1.2 (0xfefd)
            Random: 5bcba592c3df351eb8cc6b8e0d073f48c9dc243bd55d18ca...
                GMT Unix Time: Oct 21, 2018 01:00:50.000000000 +03
                Random Bytes: c3df351eb8cc6b8e0d073f48c9dc243bd55d18ca0bd6f340...
            Session ID Length: 32
            Session ID: e62dfcfdaa7222c7961f0188ff0c7f6794626b8e1947141d...
            Cipher Suite: TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256 (0xc02b)
            Compression Method: null (0)
            Extensions Length: 20
            Extension: renegotiation_info (len=1)
                Type: renegotiation_info (65281)
                Length: 1
                Renegotiation Info extension
                    Renegotiation info extension length: 0
            Extension: use_srtp (len=5)
                Type: use_srtp (14)
                Length: 5
                SRTP Protection Profiles Length: 2
                SRTP Protection Profile: SRTP_AES128_CM_HMAC_SHA1_80 (0x0001)
                MKI Length: 0
            Extension: ec_point_formats (len=2)
                Type: ec_point_formats (11)
                Length: 2
                EC point formats Length: 1
                Elliptic curves point formats (1)
                    EC point format: uncompressed (0)
    DTLSv1.2 Record Layer: Handshake Protocol: Certificate
        Content Type: Handshake (22)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 0
        Sequence Number: 1
        Length: 300
        Handshake Protocol: Certificate
            Handshake Type: Certificate (11)
            Length: 288
            Message Sequence: 1
            Fragment Offset: 0
            Fragment Length: 288
            Certificates Length: 285
            Certificates (285 bytes)
                Certificate Length: 282
                Certificate: 308201163081bda003020102020900f2ee885aa8b2553e30... (id-at-commonName=WebRTC)
                    signedCertificate
                        version: v3 (2)
                        serialNumber: 17505078724594390334
                        signature (ecdsa-with-SHA256)
                            Algorithm Id: 1.2.840.10045.4.3.2 (ecdsa-with-SHA256)
                        issuer: rdnSequence (0)
                            rdnSequence: 1 item (id-at-commonName=WebRTC)
                                RDNSequence item: 1 item (id-at-commonName=WebRTC)
                                    RelativeDistinguishedName item (id-at-commonName=WebRTC)
                                        Id: 2.5.4.3 (id-at-commonName)
                                        DirectoryString: uTF8String (4)
                                            uTF8String: WebRTC
                        validity
                            notBefore: utcTime (0)
                                utcTime: 18-10-19 22:00:50 (UTC)
                            notAfter: utcTime (0)
                                utcTime: 18-11-19 22:00:50 (UTC)
                        subject: rdnSequence (0)
                            rdnSequence: 1 item (id-at-commonName=WebRTC)
                                RDNSequence item: 1 item (id-at-commonName=WebRTC)
                                    RelativeDistinguishedName item (id-at-commonName=WebRTC)
                                        Id: 2.5.4.3 (id-at-commonName)
                                        DirectoryString: uTF8String (4)
                                            uTF8String: WebRTC
                        subjectPublicKeyInfo
                            algorithm (id-ecPublicKey)
                                Algorithm Id: 1.2.840.10045.2.1 (id-ecPublicKey)
                                ECParameters: namedCurve (0)
                                    namedCurve: 1.2.840.10045.3.1.7 (secp256r1)
                            Padding: 0
                            subjectPublicKey: 04cd8b8fb202396c7877cd55016f512083a74dfb31202625...
                    algorithmIdentifier (ecdsa-with-SHA256)
                        Algorithm Id: 1.2.840.10045.4.3.2 (ecdsa-with-SHA256)
                    Padding: 0
                    encrypted: 304502210089d194a0c6f738b054a04c9ec6c13cf3e3353e...
    DTLSv1.2 Record Layer: Handshake Protocol: Server Key Exchange
        Content Type: Handshake (22)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 0
        Sequence Number: 2
        Length: 156
        Handshake Protocol: Server Key Exchange
            Handshake Type: Server Key Exchange (12)
            Length: 144
            Message Sequence: 2
            Fragment Offset: 0
            Fragment Length: 144
            EC Diffie-Hellman Server Params
                Curve Type: named_curve (0x03)
                Named Curve: secp256r1 (0x0017)
                Pubkey Length: 65
                Pubkey: 043954b89002ed7b7b861ac638d41e2d90d783049aadfb61...
                Signature Algorithm: ecdsa_secp256r1_sha256 (0x0403)
                    Signature Hash Algorithm Hash: SHA256 (4)
                    Signature Hash Algorithm Signature: ECDSA (3)
                Signature Length: 71c
                Signature: 3045022022723b2e10f7c0ba6fb57a544a2a64415ca006ef...
    DTLSv1.2 Record Layer: Handshake Protocol: Certificate Request
        Content Type: Handshake (22)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 0
        Sequence Number: 3
        Length: 37
        Handshake Protocol: Certificate Request
            Handshake Type: Certificate Request (13)
            Length: 25
            Message Sequence: 3
            Fragment Offset: 0
            Fragment Length: 25
            Certificate types count: 2
            Certificate types (2 types)
                Certificate type: RSA Sign (1)
                Certificate type: ECDSA Sign (64)
            Signature Hash Algorithms Length: 18
            Signature Hash Algorithms (9 algorithms)
                Signature Algorithm: ecdsa_secp256r1_sha256 (0x0403)
                    Signature Hash Algorithm Hash: SHA256 (4)
                    Signature Hash Algorithm Signature: ECDSA (3)
                Signature Algorithm: rsa_pss_rsae_sha256 (0x0804)
                    Signature Hash Algorithm Hash: Unknown (8)
                    Signature Hash Algorithm Signature: Unknown (4)
                Signature Algorithm: rsa_pkcs1_sha256 (0x0401)
                    Signature Hash Algorithm Hash: SHA256 (4)
                    Signature Hash Algorithm Signature: RSA (1)
                Signature Algorithm: ecdsa_secp384r1_sha384 (0x0503)
                    Signature Hash Algorithm Hash: SHA384 (5)
                    Signature Hash Algorithm Signature: ECDSA (3)
                Signature Algorithm: rsa_pss_rsae_sha384 (0x0805)
                    Signature Hash Algorithm Hash: Unknown (8)
                    Signature Hash Algorithm Signature: Unknown (5)
                Signature Algorithm: rsa_pkcs1_sha384 (0x0501)
                    Signature Hash Algorithm Hash: SHA384 (5)
                    Signature Hash Algorithm Signature: RSA (1)
                Signature Algorithm: rsa_pss_rsae_sha512 (0x0806)
                    Signature Hash Algorithm Hash: Unknown (8)
                    Signature Hash Algorithm Signature: Unknown (6)
                Signature Algorithm: rsa_pkcs1_sha512 (0x0601)
                    Signature Hash Algorithm Hash: SHA512 (6)
                    Signature Hash Algorithm Signature: RSA (1)
                Signature Algorithm: rsa_pkcs1_sha1 (0x0201)
                    Signature Hash Algorithm Hash: SHA1 (2)
                    Signature Hash Algorithm Signature: RSA (1)
            Distinguished Names Length: 0
    DTLSv1.2 Record Layer: Handshake Protocol: Server Hello Done
        Content Type: Handshake (22)
        Version: DTLS 1.2 (0xfefd)
        Epoch: 0
        Sequence Number: 4
        Length: 12
        Handshake Protocol: Server Hello Done
            Handshake Type: Server Hello Done (14)
            Length: 0
            Message Sequence: 4
            Fragment Offset: 0
            Fragment Length: 0
```

From my debugging session it looks like presence of `rsa_pss_rsae_sha256` causes `bc-java` to fail in runtime.
I'm not sure what is proper fix, need to investigate a big more, but it sounds like unsupported algorithm should be just skipped here and not abort `DTLS` handshake.

**Comment by peterdettman** [comment]
Thanks for the bug report. Indeed it is the rsa_pss algorithms that trigger the problem, but the bug is ours, as we should be tolerating (ignoring) unrecognised hash algorithms; both RFC 8422 and 8446 introduced signature algorithms (into TLS 1.2) with the new HashAlgorithm.Intrinsic value (8).

I've pushed a fix to our git which should mirror here shortly.

On a side note, please consider updating to the new TLS API (bctls jar, org.bouncycastle.tls package). The old API in org.bouncycastle.crypto.tls is bug-fix only.

**Comment by mstyura** [comment]
Thank you very much @peterdettman. I'll try to migrate `Jitsi` to new package. I was really confused from source code, why do I have same classes in different packages, but with slightly different signatures/naming - so one of them was deprecated :).

**Comment by tylersimonmc** [comment]
I am getting this issue talking to servers using native TLS1.2 on OpenJDK11+

Do you have an ETA for a release with this fix in it?

**Comment by bcgit** [comment]
It's in the current beta at https://www.bouncycastle.org/betas if you want to try it.

**Comment by tylersimonmc** [comment]
Thanks. Works perfectly.

Any idea when the next release will be?

**Comment by bcgit** [comment]
Soon... we're aiming for before Christmas. As always it's a question of resources.

**Comment by bbaldino** [comment]
Is this fixed in v1.61?

**Comment by peterdettman** [comment]
Yes, the fix is in 1.61.

**Comment by bbaldino** [comment]
Great, thanks.  Will it make it to maven soon as wel?
